drivers/clock_control/clock_control_npcm.c

```diff
@@ -0,0 +1,365 @@
+}
+
+/* Clock controller local functions */
+static inline int npcm_clock_control_on(const struct device *dev, clock_control_subsys_t sub_system)
+{
+	uint32_t clk_id = (uint32_t)sub_system;
+	struct clk_cfg_t *priv;
+	const uint32_t pmc_base = DRV_CONFIG(dev)->base_pmc;
+
+	priv = npcm_get_cfg(sub_system);
+	if (!priv) {
+		LOG_ERR("Unsupported clock id %d", clk_id);
+		return -EINVAL;
+	}
+
+	/* Clear related PD (Power-Down) bit of module to turn on clock */
+	NPCM_PWDWN_CTL(pmc_base, NPCM_CLOCK_REG_OFFSET(priv->clock_id)) &=
+		~(BIT(NPCM_CLOCK_REG_BIT_OFFSET(priv->clock_id)));
+	return 0;
+}
+
+static inline int npcm_clock_control_off(const struct device *dev,
+					 clock_control_subsys_t sub_system)
+{
+	uint32_t clk_id = (uint32_t)sub_system;
+	struct clk_cfg_t *priv;
+	const uint32_t pmc_base = DRV_CONFIG(dev)->base_pmc;
+
+	priv = npcm_get_cfg(sub_system);
+	if (!priv) {
+		LOG_ERR("Unsupported clock id %d", clk_id);
+		return -EINVAL;
+	}
+
+	/* Set related PD (Power-Down) bit of module to turn off clock */
+	NPCM_PWDWN_CTL(pmc_base, NPCM_CLOCK_REG_OFFSET(priv->clock_id)) |=
+		~(BIT(NPCM_CLOCK_REG_BIT_OFFSET(priv->clock_id)));
+	return 0;
+}
+
+static int npcm_clock_control_get_subsys_rate(const struct device *dev,
+					      clock_control_subsys_t sub_system, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	uint32_t clk_id = (uint32_t)sub_system;
+	struct clk_cfg_t *priv;
+
+	priv = npcm_get_cfg(sub_system);
+	if (!priv) {
+		LOG_ERR("Unsupported clock id %d", clk_id);
+		return -EINVAL;
+	}
+
+	switch (priv->bus) {
+	case NPCM_CLOCK_BUS_APB1:
+		*rate = NPCM_APB_CLOCK(1);
+		break;
+	case NPCM_CLOCK_BUS_APB2:
+		*rate = NPCM_APB_CLOCK(2);
+		break;
+	case NPCM_CLOCK_BUS_APB3:
+		*rate = NPCM_APB_CLOCK(3);
+		break;
+	case NPCM_CLOCK_BUS_AHB6:
+		*rate = CORE_CLK / (AHB6DIV_VAL + 1);
+		break;
+	case NPCM_CLOCK_BUS_FIU:
+		*rate = CORE_CLK / (FIUDIV_VAL + 1);
+		break;
+	case NPCM_CLOCK_BUS_I3C:
+		*rate = CORE_CLK / (I3CDIV_VAL + 1);
+		break;
+	case NPCM_CLOCK_BUS_CORE:
+		*rate = CORE_CLK;
+		break;
+	case NPCM_CLOCK_BUS_LFCLK:
+		*rate = LFCLK;
+		break;
+	case NPCM_CLOCK_BUS_FMCLK:
+		*rate = FMCLK;
+		break;
+	case NPCM_CLOCK_BUS_USB20_CLK:
+		*rate = USB20_CLK;
+		break;
+	case NPCM_CLOCK_BUS_SIO_CLK:
+		*rate = SIO_CLK;
+		break;
+	default:
+		*rate = 0U;
+		/* Invalid parameters */
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/* Clock controller driver registration */
+static struct clock_control_driver_api npcm_clock_control_api = {
+	.on = npcm_clock_control_on,
+	.off = npcm_clock_control_off,
+	.get_rate = npcm_clock_control_get_subsys_rate,
+};
+
+static int npcm_clock_control_init(const struct device *dev)
+{
+	struct cdcg_reg *const priv = (struct cdcg_reg *)(DRV_CONFIG(dev)->base_cdcg);
+	struct freq_multiplier_t *freq_p;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(freq_multiplier); i++) {
+		if (freq_multiplier[i].ofmclk == OFMCLK) {
+			freq_p = &freq_multiplier[i];
+			break;
+		}
+	}
+
+	if (i == ARRAY_SIZE(freq_multiplier)) {
```

Comment:
I would use `>=`  not good reason feels more "defensive"

Reply:
Sure, it should cover all cases. Fixed in commit [81825c1](https://github.com/zephyrproject-rtos/zephyr/pull/78907/commits/81825c1e90f29449815aafc216c8cb5e5a7ac48d).